swift/apple/FirezoneKit/Sources/FirezoneKit/Managers/VPNConfigurationManager.swift

```diff
@@ -292,72 +295,71 @@ public class VPNConfigurationManager {
       options.merge(["id": id as NSObject]) { _, new in new }
     }
 
-    try session()?.startTunnel(options: options)
+    try session().startTunnel(options: options)
   }
 
-  func stop(clearToken: Bool = false) {
-    if clearToken {
-      do {
-        try session()?.sendProviderMessage(encoder.encode(TunnelMessage.signOut)) { _ in
-          self.session()?.stopTunnel()
-        }
-      } catch {
-        Log.error(error)
-      }
-    } else {
-      session()?.stopTunnel()
-    }
+  func signOut() throws {
+    try session([.connected, .connecting, .reasserting]).stopTunnel()
+    try session().sendProviderMessage(encoder.encode(TunnelMessage.signOut))
   }
 
-  func updateInternetResourceState() {
-    guard session()?.status == .connected else { return }
+  func stop() throws {
+    try session([.connected, .connecting, .reasserting]).stopTunnel()
+  }
 
-    try? session()?.sendProviderMessage(encoder.encode(TunnelMessage.internetResourceEnabled(internetResourceEnabled)))
+  func updateInternetResourceState() throws {
+    try session([.connected]).sendProviderMessage(
+      encoder.encode(TunnelMessage.internetResourceEnabled(internetResourceEnabled)))
   }
 
-  func toggleInternetResource(enabled: Bool) {
+  func toggleInternetResource(enabled: Bool) throws {
     internetResourceEnabled = enabled
-    updateInternetResourceState()
+    try updateInternetResourceState()
   }
 
-  func fetchResources(callback: @escaping @MainActor (ResourceList) -> Void) {
-    guard session()?.status == .connected else { return }
+  func fetchResources() async throws -> ResourceList {
+    return try await withCheckedThrowingContinuation { continuation in
+      do {
+        // Request list of resources from the provider. We send the hash of the resource list we already have.
+        // If it differs, we'll get the full list in the callback. If not, we'll get nil.
+        try session([.connected]).sendProviderMessage(
+          encoder.encode(TunnelMessage.getResourceList(resourceListHash))) { data in
 
-    do {
-      try session()?.sendProviderMessage(encoder.encode(TunnelMessage.getResourceList(resourceListHash))) { data in
-        if let data = data {
+          guard let data = data
+          else {
+            // No data returned; Resources haven't changed
+            continuation.resume(returning: self.resourcesListCache)
+
+            return
+          }
+
+          // Save hash to compare against
           self.resourceListHash = Data(SHA256.hash(data: data))
+
           let decoder = JSONDecoder()
           decoder.keyDecodingStrategy = .convertFromSnakeCase
 
-          guard let decoded = try? decoder.decode([Resource].self, from: data)
-          else {
-            fatalError("Should be able to decode ResourceList")
-          }
+          do {
```

Comment:
This nested `do` block seems redundant.

Reply:
They're both inside a closure that's not supposed to `throw`, so we have to catch errors before the closure end.